google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +400,54 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        outstandingReceipts.put(
+            message.getAckId(), new ReceiptCompleteData(outstandingMessage, false));
+      } else {
+        outstandingBatch.add(outstandingMessage);
+      }
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  synchronized void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      ReceiptCompleteData receiptCompleteData = outstandingReceipts.get(ackRequestData.getAckId());
+      OutstandingMessage outstandingMessage = receiptCompleteData.getOutstandingMessage();
+      // Setting to true means that the receipt is complete
+      receiptCompleteData.setReceiptComplete(true);
+      outstandingReceipts.put(ackRequestData.getAckId(), receiptCompleteData);
+      List<OutstandingMessage> completedReceipts = new ArrayList<>();
+      if (pendingMessages.putIfAbsent(
+              outstandingMessage.receivedMessage.getAckId(), outstandingMessage.ackHandler)
+          == null) {
+        ConcurrentHashMap<String, ReceiptCompleteData> outstandingReceiptsCopy =
+            outstandingReceipts;
+
+        for (Map.Entry<String, ReceiptCompleteData> receipts : outstandingReceiptsCopy.entrySet()) {
```

Comment:
A regular HashMap isn't going to keep track of insertion order. We should be using a LinkedHashMap to do this. We also don't need to create a copy of outstandingReceipts, we can simplify this code a lot. We should basically just pop messages from the front of the queue until receiptComplete is false.

Reply:
I have changed outstandingReceipts to be a linkedhashmap, so the order should now be guaranteed while iterating thru it. There is no pop() function in LinkedHashmap, so I am going to replace this with Iterator() so that I can safely remove elements while iterating over it.